### ISSUE TITLE
[folding] Incorrect indent level being folded

### ISSUE THREAD

**Issue by Tyriar** [Microsoft/vscode#3425]
#2994

With this Java file:

```java
package com.growingwiththeweb.sorting;

public class BubbleSort {
    public static void sort(Integer[] array) {
        for (int i = 0; i < array.length - 1; i++) {
            for (int j = 1; j < array.length - i; j++) {
                if (array[j - 1] > array[j]) {
                    swap(array, j, j - 1);
                }
            }
        }
    }

    private static void swap(Integer[] array, int a, int b) {
        Integer temp = array[a];
        array[a] = array[b];
        array[b] = temp;
    }
}
```

Selecting line number 8 (`swap(array, j, j - 1)`) and hitting <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>[</kbd> folds the wrong level:

![image](https://cloud.githubusercontent.com/assets/2193314/13299243/26bcb66e-daef-11e5-9263-b274a81787e2.png)

Adding another if statement within that `if` again folds 1 level above the correct level.

**Comment by tht13** [comment]
A further issue here is that if you perform the fold it will not fold, for example:

```javascript
function foo() {
    bar();
}
```

However if you add a second line like this:
```javascript
function foo() {
    bar();
    fux();
}
```

If you fold/unfold and then remove the new second line and attempt to fold the single line, the folding works. (The same happens with your example above, add a second line in that if statement, fold/unfold it, remove the line, it will then fold correctly.

This is caused by [`this.decorations`](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/folding/browser/folding.ts#L367) returns a two line range for single line collapse regions, so for line 8 (7, zero counting) where you trigger the fold, it will return `range: {startLine: 7, endLine: 8}` rather than `range: {startLine: 7, endLine: 7}`. 
The added line in the return range then causes the if statement in [L374](https://github.com/Microsoft/vscode/blob/master/src/vs/editor/contrib/folding/browser/folding.ts#L374) to fail and in you case it reverts to the surrounding fold range. Or in my example simply does not fold.

I have been unable to debug how these ranges are generated but I am sure @aeschli knows :smile: 

Hope this helps :+1:

**Comment by aeschli** [comment]
duplicate of #3398 

I tested the fix of #3398 with your examples. Thanks @tht13 and @Tyriar !
